Feature/add support for platform event

anmolgkv opened the pull request.

aidan-harding commented on force-app/main/default/classes/PerformanceScenario.cls:
```diff
@@ -10,4 +10,5 @@ public interface PerformanceScenario {
     void teardown();
-}
\ No newline at end of file
+    String getLabel();
```
I would make this a separate interface. Then, if a scenario doesn't declare a label, use the class name as a default. That makes it simpler to make a quick + dirty experiment

aidan-harding commented on force-app/main/default/classes/PerformanceExperiment.cls:
```diff
@@ -10,27 +10,45 @@ public with sharing class PerformanceExperiment implements Queueable {
-    public String experimentName {get; private set;}
-    public Integer thisStep {get; private set;}
+    public String suiteName { get; private set; }
+    public String experimentName { get; private set; }
```
`experimentName` and the label coming from calling `getLabel()` on the scenario are the same concept. So, only one needs to exist. 

I'd say that we can get rid of `experimentName`. That field comes all the way from the suite, so it will be a nice amount of code that we can remove here.

aidan-harding commented on force-app/main/default/classes/PerformanceExperiment.cls:
This class has grown into a quite a beast. It as a bit of a beast before, but it seems to have reached a tipping point for me now. 

We can simplify with a couple of things:

1. Remove getClassNameOf() and put it in a separate class.
2. Create a builder class as an inner class. I originally wrote a builder-style version of this class but without creating an actual builder class. Now, with the JSON deserialise part, a lot of the class is just about construction. So, by adding an inner class called Builder, all the build-stuff is grouped together and away from the actual execution.

aidan-harding commented on force-app/main/default/classes/PerformanceExperimentEvent.cls:
```diff
@@ -0,0 +1,35 @@
+
+            try {
+                thisExperiment.run();
+            } catch (Exception ex) {
```
Why catch and then rethrow?

aidan-harding commented on force-app/main/default/classes/PerformanceScenarioRunner.cls:
```diff
@@ -4,19 +4,28 @@
+    private PerformanceTestingMode mode;
 
-    public PerformanceScenarioRunner(PerformanceScenario performanceScenario) {
+    public PerformanceScenarioRunner(PerformanceScenario performanceScenario, PerformanceTestingMode mode) {
```
Would it make more sense to pass in the measure rather than the mode here? The runner doesn't really need to know that PEs don't support CPU time properly. It should just measure what it's told to measure.

aidan-harding commented on force-app/main/default/classes/PerformanceSuiteResponse.cls:
Can you please add a comment header to the file, summarising the purpose of the class?

aidan-harding commented on force-app/main/default/classes/PerformanceSuiteRunner.cls:
Can you add an example of the PerformanceSuiteRunner? I gather it's how the CLI runs a suite, but there's not example in the repo

aidan-harding reviewed (changes requested):
Hi Anmol,

This is a good stuff. There are quite a few comments from my review - partly because this will be open source so it's good to be on best behaviour. Feel free to push back or ask questions...

anmolgkv commented:
> Hi Anmol,
> 
> This is a good stuff. There are quite a few comments from my review - partly because this will be open source so it's good to be on best behaviour. Feel free to push back or ask questions...

Thank you @aidan-harding for the review, I have cleaned up the code a bit.
Can I remove the gnuplot script from repo and update readme with reference to the cli extension?
I have also changed the schema a bit(Removed CPUTime field and added a generic Performance Metric field to capture: `CPU Time`, `Clock Time` and `Heap size`) 

The GNU plot will need modification if we are going to keep it in the repo

aidan-harding commented:
> > Hi Anmol,
> > This is a good stuff. There are quite a few comments from my review - partly because this will be open source so it's good to be on best behaviour. Feel free to push back or ask questions...
> 
> Thank you @aidan-harding for the review, I have cleaned up the code a bit. Can I remove the gnuplot script from repo and update readme with reference to the cli extension? I have also changed the schema a bit(Removed CPUTime field and added a generic Performance Metric field to capture: `CPU Time`, `Clock Time` and `Heap size`)
> 
> The GNU plot will need modification if we are going to keep it in the repo

Hi Anmol,

Yes, go ahead and remove gnuplot. The CLI version is much easier and neater. 

I'm not sure about changing the field. At the moment, we only record one metric per scenario. I think it would be a good future extension to be able to record multiple metrics from a single suite run. For example, elapsed time vs. CPU time is interesting to measure together. It shows the cost of DB operations and trade-offs e.g. we think that Audicity in PE mode is faster on the clock than Queueable, but the same for CPU time. And a graph of that would be handy. 

I'm not saying that we need to add the ability to record multiple dimensions today. But, let's keep that option open by keeping the fields on PerformanceMeasureResult separate.

anmolgkv commented:
> > > Hi Anmol,
> > > This is a good stuff. There are quite a few comments from my review - partly because this will be open source so it's good to be on best behaviour. Feel free to push back or ask questions...
> > 
> > 
> > Thank you @aidan-harding for the review, I have cleaned up the code a bit. Can I remove the gnuplot script from repo and update readme with reference to the cli extension? I have also changed the schema a bit(Removed CPUTime field and added a generic Performance Metric field to capture: `CPU Time`, `Clock Time` and `Heap size`)
> > The GNU plot will need modification if we are going to keep it in the repo
> 
> Hi Anmol,
> 
> Yes, go ahead and remove gnuplot. The CLI version is much easier and neater.
> 
> I'm not sure about changing the field. At the moment, we only record one metric per scenario. I think it would be a good future extension to be able to record multiple metrics from a single suite run. For example, elapsed time vs. CPU time is interesting to measure together. It shows the cost of DB operations and trade-offs e.g. we think that Audicity in PE mode is faster on the clock than Queueable, but the same for CPU time. And a graph of that would be handy.
> 
> I'm not saying that we need to add the ability to record multiple dimensions today. But, let's keep that option open by keeping the fields on PerformanceMeasureResult separate.

I am capturing all this in the json field `{"realTime":10.0,"heapSize":19.0,"cpuTime":6}` and from cli user has option to generate report on any/all of the metrics. so I was thinking the CPU Time field is redundant, or did you mean something else?

aidan-harding commented:
> I am capturing all this in the json field `{"realTime":10.0,"heapSize":19.0,"cpuTime":6}` and from cli user has option to generate report on any/all of the metrics. so I was thinking the CPU Time field is redundant, or did you mean something else?

I hadn't noticed that they were going into JSON. I think it would be better to go into top-level fields unless there's a reason I'm missing for wanting JSON?

anmolgkv commented:
> > I am capturing all this in the json field `{"realTime":10.0,"heapSize":19.0,"cpuTime":6}` and from cli user has option to generate report on any/all of the metrics. so I was thinking the CPU Time field is redundant, or did you mean something else?
> 
> I hadn't noticed that they were going into JSON. I think it would be better to go into top-level fields unless there's a reason I'm missing for wanting JSON?

Just for extensibility, in future if we want to capture any other metrics, but happy to move it to top level methods

aidan-harding commented:
> > > I am capturing all this in the json field `{"realTime":10.0,"heapSize":19.0,"cpuTime":6}` and from cli user has option to generate report on any/all of the metrics. so I was thinking the CPU Time field is redundant, or did you mean something else?
> > 
> > 
> > I hadn't noticed that they were going into JSON. I think it would be better to go into top-level fields unless there's a reason I'm missing for wanting JSON?
> 
> Just for extensibility, in future if we want to capture any other metrics, but happy to move it to top level methods

OK, thanks. I think that since the number of metrics is never going to be huge, and since this package is unlocked, then the top-level fields are a better fit. It allows the data to be manipulated more easily with other tools

anmolgkv commented:
@aidan-harding I have done all the changes please review

aidan-harding commented on README.md:
```diff
@@ -163,29 +162,89 @@ So we can conclude that for performance-critical code, this is the best method t
+There are 3 ways you can do performance Testing with this library:
+1. **QUEUEABLE**: Runs each performance test scenario in queueable mode, and chain the next execution in finalizer
+2. **PLATFORM_EVENTS**: Runs each performance test scenario in synchronous mode, and chain the next execution by firing platform event
+3. **QUEUEABLE_FROM_FINALIZER**: Runs each performance test scenario in queueable mode(from finalizer class), and chain the next execution in finalizer
```
It would be good to include a paragraph on how to choose between those three modes.

aidan-harding reviewed (approved):
This looks good, thanks @anmolgkv. I've asked for one extra bit in the README, but otherwise it's good to go!
